### DISCUSSION TITLE
Shadow Offset (ground-position attribute PRs welcome)

### DISCUSSION THREAD

**Discussion by Lemovision** [google/model-viewer#3985] [General]
![shadowOffset](https://user-images.githubusercontent.com/61175245/205511214-1a549489-f5ca-453f-8dcc-a3babfa26116.png)

https://user-images.githubusercontent.com/61175245/205512285-ed84333f-686b-4ac7-b020-b96472f884aa.mp4

https://www.elephantparade.com/view/?e=5 (click walk button)

I noticed my model appears to be floating slightly above the ground when activating shadows
Actually in my animation the feet perfectly touch the ground plane at the origin,
expect a few frames with the heel that slightly clips into the ground,
but that's perfectly fine usually for a walk animation as the feet would bend in real world
or just disappear into the ground in a normal 3d scene

So it seems your shadows might be created at the lowest point in the model with animation?
Which I think doesn't always make sense depending of the animation
Is it possible to have the shadow created at ground/origin, as intended in the 3d model file?
Or maybe a separate shadow-position attribute? -> "minimum"(default) or "origin" (world coordinates zero)

**Comment by elalish** [comment]
We definitely don't use the lowest point; we don't scan the whole animation at all. I believe it's based on the resting pose. Could it be that your resting pose is offset, or at that foot-angled pose? I'd actually love to know what exactly the artist steps are to resolve this kind of problem; it'll be helpful for others, but also clue me in to what common flows look like and how we can improve the experience.

**Reply by elalish** [reply]
Indeed, I can actually see you model move vertically when starting or stopping the walking animation, so I think that's it.

**Reply by Lemovision** [reply]
![offset](https://user-images.githubusercontent.com/61175245/207149897-deb88ce6-8ad1-4fc9-9336-2242458f1315.gif)
There is indeed a slight position difference between the walking pose and the rest pose,
But as you can see in this GIF, they're still both under the green ground line in my 3d software (blender here)
Actually the rest pose is even quite a bit too much into the ground
But that's not what I'm seeing when displayed in model-viewer, where even the rest pose is mostly floating,
except for the lowest foot, which seem to be on the ground

Could it be then (regardless of animation) the lowest point of the rest pose that drives your ground/shadow estimation?
Or are you saying there's no ground estimation and ground supposed to be same as in model's world coordinates (green line)?

PS: It is possible there are some exporting problem on my side because I'm using Maya for animation,
But I have to pass it to Blender via FBX for the better GLB export support, and I'm not totally familiar with Blender
(And who know maybe there's some conversion offset somewhere in between)

**Reply by elalish** [reply]
Indeed, we use the lowest point of the rest post as the ground. We don't trust the absolute model coordinates to mean anything because they are so often wrong.

**Reply by Lemovision** [reply]
Okay thanks for the confirmation
With this explanation I will be able to adjust my model and compensate for that for now

In my opinion though, most people designing in 3d programs would use the ground as a reference,
and would expect a "what you see is what you get" workflow with the same ground in model-viewer
That's why I found it very confusing that my model was floating, although I didn't place it like that
Also like I said it looks more natural in some animations to have slight floor clipping rather than floating
People also might want to levitate or fly objects above the ground, which might not work with "lowest point"
I've seen these other threads with people being confused by that system

https://github.com/google/model-viewer/discussions/1872
https://github.com/google/model-viewer/discussions/2204
https://github.com/google/model-viewer/issues/998

So as I suggested it would be great I think to have a **"shadow-position"** or **"ground-position"** attribute
The default value is "lowest" as currently, and another "origin" option to keep model world coordinates
Like this there's no need for awkward "offset translation" configurations
And it's possible to place the model in the 3d  program above or under the ground visually as designed
So that it would appear in the same spot in model-viewer when **ground-position="origin"**

**Reply by elalish** [reply]
I like the idea of `ground-position="origin"` as an API. How would that interact with `getDimensions` and the default `camera-target`? PRs welcome.

**Reply by Lemovision** [reply]
Okay glad you like the idea
If anyone manages to add this it would be great
